packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol

```diff
@@ -263,20 +271,34 @@ abstract contract CrossDomainMessenger is
 
         require(successfulMessages[versionedHash] == false, "CrossDomainMessenger: message has already been relayed");
 
+        // Check the relay message validator configuration. On L1, this MUST be a no-op and return the zero address.
+        // On L2, this CAN perform a static call to get the message validator config.
+        address messageValidator = _relayMessageValidatorConfig();
+
         // If there is not enough gas left to perform the external call and finish the execution,
         // return early and assign the message to the failedMessages mapping.
         // We are asserting that we have enough gas to:
-        // 1. Call the target contract (_minGasLimit + RELAY_CALL_OVERHEAD + RELAY_GAS_CHECK_BUFFER)
+        // 1. Call the target contract (_minGasLimit + RELAY_CALL_OVERHEAD + RELAY_GAS_CHECK_BUFFER +
+        // _relayMessageValidatorConfigGas)
         //   1.a. The RELAY_CALL_OVERHEAD is included in `hasMinGas`.
+        //   1.b. The relay message validation check is included in `hasMinGas`.
         // 2. Finish the execution after the external call (RELAY_RESERVED_GAS).
         //
         // If `xDomainMsgSender` is not the default L2 sender, this function
         // is being re-entered. This marks the message as failed to allow it to be replayed.
+        //
+        // If relay message validation fails (only possible on L2), then mark the message as failed to allow
+        // it to be replayed.
+        // TODO: Do we want the estimation address check and reversion after failed message validation? This
+        // can be a valid UX flow for getting messages from L1 -> L2 that are later replayed, we don't want
+        // to confuse users from sending L1 -> L2 messages that are meant to be replayed.
         if (
             !SafeCall.hasMinGas(
                 _minGasLimit,
-                RELAY_RESERVED_GAS + RELAY_GAS_CHECK_BUFFER + _relayMessageValidationGas(uint64(_message.length))
+                RELAY_RESERVED_GAS + RELAY_GAS_CHECK_BUFFER
+                    + _relayMessageValidatorGas(messageValidator, uint64(_message.length))
             ) || xDomainMsgSender != Constants.DEFAULT_L2_SENDER
+                || !_isRelayMessageValidated(messageValidator, _nonce, _sender, _target, _value, _message)
```

Comment:
`!_isRelayMessageValidated(messageValidator, _nonce, _sender, _target, _value, _message)` This should be after `failedMessages[versionHash]` right?

Reply:
the SafeCall.hasMinGas will fail first if we don't have enough gas to do the `_isRelayMessageValidated` call

Reply:
Yeah that's what I'm asking. So it will break out of the or clause the second the `hasMinGas` returns false?

Reply:
yea:
- https://stackoverflow.com/a/42967711
- https://docs.soliditylang.org/en/latest/types.html#booleans
- https://docs.soliditylang.org/en/latest/control-structures.html#order-of-evaluation-of-expressions